docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,230 @@
+
+We could attempt to get around this either by heavily limiting the language in type functions to avoid Turing-completeness or by introducing heuristics to try to detect infinite loops while the type function is running. We propose doing neither. In the former case, the restrictions to the language would be significant, heavily limiting the use of while loops, function calls, etc. This would mean that learning to program type functions would feel _very_ different from learning to program Luau in general. In the latter case, the implementation work is considerable, and it's not clear what kind of impact it might have on the VM itself. 
+
+Instead, this RFC proposes that we accept that type functions, like the rest of Luau's type system, are already not guaranteed to terminate in general, and we should employ the same mitigations that we've already implemented for addressing that problem --- namely, Luau's analysis supports global limits that include an overall runtime limit on analysis and a user-requestable (really, embedder-requestable) cancellation system for analysis. This means that editors, language services, and other software embedding Luau's type analysis will continue to be able to use one simple, unified system for imposing limits on analysis. In an editor tooling context, we expect that it might be useful for us to grow the feedback provided by the cancellation to support editors providing more detailed feedback to the users. It's not hard to imagine a developer experience where your editor informs you that a particular type function call in a particular spot in your program is taking a long time to evaluate. This would actually be a better experience than what is available today when builtin pieces of Luau's type system are unable to complete in an acceptable amount of time, and we are unable to give any actionable feedback to the user. Extending such a system to support identifying problem areas in built-in components of the type system is likely a much harder effort that is out of scope for this RFC.
+
+### `types` API Reference
+
+This section details the initial programming interface we propose for `type`s when they are reflected into type function bodies. Each section is separated by headers (e.g. Singleton) and describe the methods available to that category of type. The initial section, titled `types`, describes the elements of the interface common to every category of type. All attributes of newly-created `type`s are initialized with empty tables / arrays and `nil`. All arguments are passed by references. 
+
+<details><summary>Expand for full type API reference.</summary>
+
+#### `types`
```

Comment:
Maybe the next 3 tables can be formatted a bit better for clarity.

The section can start with "`types` library" where "Instance Attributes" are replaced with "Library Properties" and "Static Methods" are replaced with "Library Functions". Functions because they are called as `types.singleton(nil)`.
I don't think there's a mention to make library itself available through the `type` (like with the `string` library) and I do not suggest doing so (so no `t:unionof(types.number)`).

With this clear separation we can then make a split for `type` heading and place "Instance Methods" there.

And before moving on to `Negation` and other section we can note that `type` can have a 'tag' or a 'kind'. By the way, we have the `is` method to check the tag/kind, but there is no property to read what the kind is? This I would suggest adding.

What I mean about the structure:

#### `types` Library

| Library Properties | Type | Description |
| ------------- | ------------- | ------------- |
| `unknown` | `type` | an immutable instance of the built-in type `unknown` |
| `string` | `type` | an immutable instance of the built-in type `string` |
...

| Library Functions | Return Type | Description |
| ------------- | ------------- | ------------- |
| `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |
...

### `type` instance

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
...

Depending on the tag, `type` instance can have additional properties and methods available.

#### Negation `type`

...

Reply:
Really appreciate this suggestion. Just reformatted it! 😸